### ISSUE TITLE
Make some functions in pl-wic.c public

### ISSUE THREAD

**Issue by kamahen** [SWI-Prolog/swipl-devel#1120]
Possibly we should make some of the low-level functions of pl-wic.c public (as PL_*). I.e., we do have the stuff for saving atoms and integers there that takes care of all issues related to that, such as encoding, big/little endian, different sizes, etc. Maybe just for atoms and integers, at least initially.

These can be tested with the save/load functions for packages/pcre.

Please assign this issue to me.

(This is in response to https://github.com/SWI-Prolog/packages-pcre/pull/27#issuecomment-1423823764)

**Comment by kamahen** [comment]
@JanWielemaker - what does "wic" mean? Is it an old name for "qlf"? 
Reason for asking: I'll want to put some kind of uniform prefix on the public functions, e.g. `PL_qlf_putString()` (or `PL_wic_putString()`).

**Comment by JanWielemaker** [comment]
Warren Intermediate Code or something like that.  As the VM is not a WAM, it is a bad name.   qlf is at least not as bad (Quick Load File/Format).   I wouldn't mind a better name.    The file was called like this from day 1 of SWI-Prolog (before it actually had a name :smile: )

**Comment by kamahen** [comment]
Also need to update foreign.doc. There is some preliminary stuff in https://github.com/SWI-Prolog/swipl-devel/pull/1122

**Comment by kamahen** [comment]
Leaving this open because not all functions have been exported

**Comment by JanWielemaker** [comment]
You probably notably want to add atoms.   That saves all the encoding issues and atoms are saved only once in .qlf files,
saving time and space.    After that, I think we can close.   If there is a serious demand we can always add more.

**Comment by kamahen** [comment]
`putAtom()` has a parameter `wic_state*` (`getAtom()` just has `IOSTREAM*`). From tracing the code, it seems that the following fields of `wic_state*` are used:
* `wicFd` (`IOSTREAM*`)
* `idMap` (table of mapped identifiers)
* `savedXRTable`

It's not clear to me how these can be disentangled without changing the "save"/"load" API.

**Comment by kamahen** [comment]
It appears that `state->idMap` can be `NULL`. However `saavedXRTable` is more problematic -- it's used (indirectly from `saveXRBlobType()` in `savedXR()`.

**Comment by kamahen** [comment]
If we don't care about optimizing the output, is there any harm in outputting an atom as a string (and not an index)? That is, does the "load" assume that each time it gets an an `XR_ATOM` that it hasn't been encountered before, or will its lookup just find the already existing index? (From a quick read of the code, it looks like the latter).

If I don't have `PL_qlf_putAtom()`, I'd just output the string as part of my "save", and then call `PL_new_atom_nchars()` or similar; as far as I can tell, my proposed `PL_qlf_putAtom()` is essentially the same.

**Comment by JanWielemaker** [comment]
Of course you can save atoms as a string, but not using the XR_ATOM code as that will get the numbering out of sync when loading.    Anyway, if we want PL_qlf_putAtom() it should also use the infrastructure.   Make PL_qlf_putAtom() do something different than what normally happens with atoms seems silly.   What if the atom is in fact a blob?    Being able to save and load atoms can be handy as it is not always simple (encoding, 0-bytes, blobs) and using the built-ins minimizes duplication.    

For saving we already have LD->qlf.current_state it seems.    For loading we pass it around.   We would need a new field and only use it around the callback as loading a qlf may trigger loading other qlf's and even creating new ones.

**Comment by kamahen** [comment]
By "save atom as string", I meant that if I had a blob that contained an `atom_t` (which I knew wasn't a blob because I'd created it), its "save" function would save the atom's contents (probably using `PL_qlf_putString()`); and in the "load" function, `PL_new_atom_nchars()` would be used with that saved string.

`PL_qlf_putAtom()` and `PL_qlf_getAtom()` would *not* be used by anything in `pl-wic.c` -- they would be strictly for use by blob "save" and "load" functions. (And, yes: they should be able to handle both text atoms and arbitrary blobs). The existing `putAtom()` and `getAtom()` would continue to use the references -- `PL_qlf_putAtom()` and `PL_glf_getAtom()` wouldn't use references -- either they'd use my proposed code to create and delete a `wic_state` or they'd be separate functions that are essentially `putAtom()` and `getAtom()` without the reference lookups that require `wic_state` [`PL_qlf_putAtom()` might be nothing more than a call to `putAtom()`].

As for your proposed "global" `wic_state` -- as a general principle, I don't like global context that doesn't show up as a parameter. Besides being "magical", it also makes writing unit tests for the "save" functions a bit harder. As an alternative to your proposed global state, perhaps add a field to `IOSTREAM` that's only for loading, and contains a pointer to the `wic_state`? This might be the simplest solution, although we'd probably also need `PL_qlf_setup_iostream()` for unit tests (which would create a `wic_state` and add it to the `IOSTREAM`) and a corresponding `PL_qlf_teardown_iostream()`.

(I presume that the reason that `getAtom()` doesn't need a `state` is that it's only called after seeing a `XR_ATOM`; that is for the first appearance of an atom.)

Anyway, my preferred solution is to hat `PL_qlf_getAtom()` that doesn't need a `wic_state`, even if that's a bit less efficient.

**Comment by JanWielemaker** [comment]
I agree that hacking with a temporary thread specific variable is not ideal.   Extending IOSTREAM is also not what I want.   Adding an argument to the save/load hooks would be the nicest, but would cause compiler errors on any code using this.    I surely do not want a PL_qlf_putAtom() that doesn't use the infrastructure of pl-wic.c.

So, I'm happy (enough) with the LD field hack and I'm happy with no atom support at all, although I think that is a lost opportunity.    Saving and loading text is also non-trivial as you have to deal with the encoding, 0-bytes, reserving enough space for loading, etc.

I don't see the test issues that much.    A save/load test with library(pcre) should be enough.    You can be fairly sure atoms can be correctly saved and loaded because if not there is a lot that does not work.   They already get a lot of testing from simply building the system.

**Comment by kamahen** [comment]
I scanned the codebase for save/load, and only found one, for `ucs_atom` (in `pl-fli.c`). As far as I can tell, all the other save/loads return an error (via `PL_message()`) from the "save" function. So, changing the signature for save/load probably won't break anything -- but also I don't think that needs to be done (see example code below).

The `XR_REF` lookup only applies to handing general terms. Inside a blob, it's up to the save/load functions to determine what happens. A proposed `PL_qlf_putAtom()` would be a wrapper on `PL_qlf_putString()`, and similarly the `PL_qlf_getAtom()` would be a wrapper on `PL_qlf_getString()`. (We probably also want an API for creating an atom blob.)
 
Here's what I think the code should be for a simplified version of the PCRE blob (untested/uncompiled code):
```cpp
static PL_blob_t pcre2_blob =
{ PL_BLOB_MAGIC,
  0,
  "regex",
  release_pcre,
  compare_pcres,
  write_pcre,
  NULL, /* acquire */
  save_pcre,
  load_pcre
};

typedef struct re_data
{ atom_t        pattern;
  uint32_t      flags;
  pcre2_code_8 *pcre2_code;
};

static int
save_pcre(atom_t symbol, IOSTREAM *fd)
{ const re_data *re = PL_blob_data(symbol, NULL, NULL);
  PL_STRINGS_MARK();
    size_t pattern_len;
    const char *pattern_str = PL_atom_nchars(re->pattern, &pattern_len);
    PL_qlf_putString(pattern_str, pattern_len, fd);
    PL_qlf_putAtom(re->pattern, fd);
    PL_qlf_putUint(re->flags, fd);
    // Don't save the pcre2_code; it'll be generated by the load
  PL_STRINGS_RELEASE();
  return TRUE;
}

static atom_t
load_pcre(IOSTREAM *fd)
{ re_data re;
  size_t pattern_len;
  const char *pattern_str = PL_qlf_getString(fd, &pattern_len);
  re.pattern = PL_new_atom_nchars(pattern_len, pattern_str);
  re.flags = PL_qlf_getUint(fd);
  re.pcre_code = pcre2_compile(pattern_str, pattern_len, re.flags, ...);
  if ( !re.pcre_code )
  { Sseterr(fd, SIO_FERR, "PCRE failed to compile");
    return;
  }
  // The following may be wrong, but it's the only way I could see
  // for creating an atom blob of type pcre2_blob:
  term_t blob_term = PL_new_term_ref();
  PL_put_blob(blob_term, &re, sizeof re, &pcre2_blob);
  atom_t a;
  PL_get_atom(blob_term, &a);
  return a;
}
```

**Comment by JanWielemaker** [comment]
> So, changing the signature for save/load probably won't break anything

Bit most do define it, so if we change the signature the compiler is going to complain.    At least with the default settings, this results in failure to compile.    There is lots of code out there that uses blobs.

> ```c++
> const char *pattern_str = PL_atom_nchars(re->pattern, &pattern_len);
> ```

That only works for ISO Latin 1 strings.    The qlf I/O for strings is wrong as is.   Needs some thought.

**Comment by kamahen** [comment]
For the real thing, I'd output wchars, which should work for everything. I just copy/pasted some existing code for this example. And I'd want to encapsulate it all in PL_qlf_{put,get}Atom() functions.

Is my way of creating a blob atom correct, or is there a better way?
